File: ex/uva100.cpp
#include <iostream>
using namespace std;

int n1, n2;

int cycle_length(int n, int cc)
{
  if (n == 1)
    return cc + 1;
  else if (n % 2 == 0)
    return cycle_length(n/2, cc+1);
  else
    return cycle_length(n*3+1, cc+1);
}

main()
{
  int result;
  int biggest;

  while (cin >> n1 >> n2)
  {
    biggest = 0;

    for (int i = n1; i < n2; ++i)
    {
      /* calculate */
      result = cycle_length(i, 0);

      /* attribuitions */
      if (result > biggest)
        biggest = result;
    }

    cout << n1 << " " << n2 << " " << biggest << endl;
  }
}

File: ex/uva100.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Map;
import java.util.HashMap;

class uva100 {
  static InputStreamReader ir = new InputStreamReader(System.in);
  static BufferedReader in = new BufferedReader(ir);

  static Map memo = new HashMap<Integer, Integer>();
  static int lowBound, highBound;

  static int cycleLength(int number, int cycle) {
    int offset = cycle;

    if (memo.containsKey(number))
      cycle += (int) memo.get(number);
    else {
      if (number % 2 != 0)
        cycle = cycleLength(number * 3 + 1, ++cycle);
      else
        cycle = cycleLength(number / 2, ++cycle);
      memo.put(number, cycle - offset);
    }

    return cycle;
  }

  static boolean readData() throws IOException {
    boolean flag = true;
    String inlet = in.readLine();

    try {
      String[] line = inlet.split(" ");
      lowBound = Integer.parseInt(line[0]);
      highBound = Integer.parseInt(line[1]);

      if (highBound < lowBound) {
        int temp = highBound;
        highBound = lowBound;
        lowBound = temp;
      }
    }
    catch (Exception NullPointerException) {
      flag = false;
    }

    return flag;
  }

  public static final void main(String[] args) throws IOException {
    int result;
    int biggest;
    memo.put(1, 1);

    while (readData()) {
      biggest = 0;

      for (int n = lowBound; n <= highBound; ++n) {
        result = cycleLength(n, 0);

        if (result > biggest)
          biggest = result;
      }

      System.out.println(lowBound + " " + highBound +  " " + biggest);
    }
  }

}
